### ISSUE TITLE
Generalised per-room memory

### ISSUE THREAD

**Issue by Cadair** [opsdroid/opsdroid#1733]
The matrix database stores data in room state. This means that the database is bound to a specific room, there is also a couple of helper functions which allow you to use memory in a specific room, this is very useful as it allows the memory to have the context of the room in which the message was received. So you can store per-chat context along with the memory.

At the moment, this is a matrix only feature, which means that skills written to use it would break pretty horrifically if the matrix database was not configured. This issue is a feature request to generalise the opsdroid Memory implementation to support per-target memory.

Replicating the behaviour of the matrix database where it defaults to a global store, but there is a context manager and a decorator which allow you to specific which target the memory is related to seems like a sensible idea. The big question though is how to implement this at the database layer. For matrix the room in which the state is stored would be how it is implemented, for sqlite you could imagine a table per target, but then for some databases you might just want to prefix the key with the target id before reading or writing to the store.

**Comment by SolarDrew** [comment]
To some extent this is already kind of covered by the `memory_in_room` context manager, right? Other databses could define this (or equivalent) to namespace entries in whatever way makes sense for that kind of memory, and without impacting how the matrix one works. I'm less sure how you'd approach defaulting to the room an event was received in.

**Comment by Cadair** [comment]
I don't think we need to default to the room in which the event was received, just all the databases need to support memory in room for the feature to not lock you into the matrix db.

What I am suggesting is that memory in room becomes part of Memory and all dbs support it. Then the memory_in_event_room works for everyone all the time.
